USB: Unifying legacy and composite (including cherry-picked)

finikorg opened the pull request:
Cherry-picked patches from PR #11041 (which was closed). Rebased against latest master.

@aurel32 please push any missing patches to this branch if needed.

aurel32 committed:
usb: merge all request handler buffers

Now that we have merged the class and vendor request buffers into the
USB device code, we can also merge the standard request buffer. We just
need to ensure it is at least 8 bytes, by using the range option in
Kconfig. The 64KiB limit is arbitrary.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
usb: move the request handler buffer to the USB device code

In order to unify the legacy and composite code, move the class and
vendor request handler buffer into the USB device code, just like in
composite mode. The option is renamed from USB_COMPOSITE_BUFFER_SIZE
into USB_REQUEST_BUFFER_SIZE and also replaces the USB_DFU_MAX_XFER_SIZE
and USB_HID_MAX_PAYLOAD_SIZE options.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>
Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

aurel32 committed:
usb: simplify request handlers code

Now that Standard, Class and Vendor request handlers are using the same
buffer, we do not need to keep and array installed request data pointer.
This simplifies the code.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>
Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Set correct default buffer size for RNDIS

Set correct USB_REQUEST_BUFFER_SIZE for USB_DEVICE_NETWORK_RNDIS.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: net: echo_server: Fix usbnet configuration

This patch enables USB tests.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: device: Remove unused variable

Remove unused variable.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jfischer-no reviewed (changes requested):
LGTM, except parts regarding default endpoint addresses (AUTO_EP_*)

finikorg committed:
usb: Align legacy and composite interface

Remove duplicated execution path for composite configuration, USB
device stack initialization is done inside stack for both cases.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: usb: device: Refactor device api test

Refactor device api test using standard placement for descriptor table
and configuration data.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: usb: webusb: Use new descriptor placement API

Simplify code using new descriptor API and remove old composite
initialization.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: usb: webusb: Use Automatic endpoint assignments

Use endpoint index since addresses are automatically assigned.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: usb: wpanusb: Refactor sample using new API

Refactor wpanusb sample using new descriptor API, use endpoint index
instead of hardcoded endpoint address.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: dfu: Remove old composite configuration

Use new method for USB config and enable.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: usb: hid: Fix harness configuration

Fix configuration indicating changes in USB device.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
doc: usb: Remove old API documentation

Fix documentation build removing old API tags.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

dbkinder reviewed (approved):
Doc changes LGTM.

jfischer-phytec-iot commented:
@finikorg Is it still WIP?

jfischer-no reviewed (changes requested):
LGTM, just few comments

finikorg commented on subsys/usb/class/netusb/netusb.c:
```diff
@@ -137,34 +137,6 @@ static void netusb_init(struct net_if *iface)
 
 	net_if_down(iface);
 
-#ifndef CONFIG_USB_COMPOSITE_DEVICE
```
done

finikorg commented on tests/subsys/usb/desc_sections/src/desc_sections.c:
```diff
@@ -221,7 +221,7 @@ static void test_desc_sections(void)
 			"USB Configuratio structures section");
 
-	head = (struct usb_desc_header *)usb_get_device_descriptor();
+	head = (struct usb_desc_header *)__usb_descriptor_start;
```
done

finikorg commented:
Rebased and fixed conflicts with master.

jfischer-phytec-iot commented:
Tested it for a while:
hci_usb :x: (on master :x:)
cdc_acm :white_check_mark: 
cdc_acm_composite :white_check_mark: 
dfu :white_check_mark: 
dfu + composite mode :x: (class_handler (and custom_handler) is hardcoded to primary set of descriptors, dfu has two set of descriptors and does not work in composite configuration)
hid-mouse :white_check_mark: 
hid-cdc :white_check_mark: 
mass :white_check_mark: 
testusb :white_check_mark:

masz-nordic commented:
Tested with `nrf52840_pca10056`:

|       Sample      |       Result       |
|:-----------------:|:------------------:|
|      cdc_acm      | :heavy_check_mark: |
| cdc_acm_composite | :heavy_check_mark: |
|        hid        | :heavy_check_mark: |
|     hid-mouse     | :heavy_check_mark: |
|      hid-cdc      | :heavy_check_mark: |
|        mass       | :heavy_check_mark: |

Although all of them exhibited reconnection issues, as expected.

jfischer-no committed:
usb: usb_device: fix handlers call for secondary descriptor

Fix handlers call for secondary descriptor.
USB class like DFU has two sets of descriptors. In the handler
look up function, wind the pointer forward to the range
of the current descriptor.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>
